std/experimental/color/conv.d

```diff
@@ -0,0 +1,523 @@
+    static assert(colorFromString("#FF8000") == RGB8(0xFF,0x80, 0x00));
+    static assert(colorFromString("$FF8000") == RGB8(0xFF,0x80, 0x00));
+    static assert(colorFromString("0xFF8000") == RGB8(0xFF,0x80, 0x00));
+
+    // 4/8 digita (/w alpha)
+    static assert(colorFromString!RGBA8("#8C41") == RGBA8(0xCC,0x44, 0x11, 0x88));
+    static assert(colorFromString!RGBA8("#80CC4401") == RGBA8(0xCC,0x44, 0x01, 0x80));
+}
+
+
+package:
+
+// convert between pixel data types
+To convertPixelType(To, From)(From v) if(isValidComponentType!From && isValidComponentType!To)
+{
+    import std.algorithm: max;
+
+    static if(isIntegral!From && isIntegral!To)
+    {
+        // extending normalised integer types is not trivial
+        return convertNormInt!To(v);
+    }
+    else static if(isIntegral!From && isFloatingPoint!To)
+    {
+        alias FP = FloatTypeFor!(From, To);
+        static if(isSigned!From) // max(c, -1) is the signed conversion followed by D3D, OpenGL, etc.
+            return To(max(v*FP(1.0/From.max), FP(-1.0)));
+        else
+            return To(v*FP(1.0/From.max));
+    }
+    else static if(isFloatingPoint!From && isIntegral!To)
+    {
+        alias FP = FloatTypeFor!(To, From);
+        // HACK: this is incomplete!
+        //       +0.5 rounding only works for positive numbers
+        //       we also need to clamp (saturate) [To.min, To.max]
+        return cast(To)(v*FP(To.max) + FP(0.5));
+    }
+    else
+        return To(v);
+}
+
+
+// converts directly between fixed-point color types, without doing float conversions
+// ** this should be tested for performance; we can optimise the small->large conversions with table lookups
+To convertNormInt(To, From)(From i) if(isIntegral!To && isIntegral!From)
+{
+    import std.traits: isUnsigned, Unsigned;
+    template Iota(alias start, alias end)
+    {
+        static if(end == start)
+            alias Iota = TypeTuple!();
+        else
+            alias Iota = TypeTuple!(Iota!(start, end-1), end-1);
+    }
+    enum Bits(T) = T.sizeof*8;
+
+    static if(isUnsigned!To && isUnsigned!From)
+    {
+        static if(Bits!To <= Bits!From)
+            return To(i >> (Bits!From-Bits!To));
+        else
+        {
+            To r;
+            enum numReps = Bits!To/Bits!From;
+            foreach(j; Iota!(0, numReps))
+                r |= To(i) << (j*Bits!From);
+            return r;
+        }
+    }
+    else static if(isUnsigned!To)
+    {
+        if(i < 0) // if i is negative, return 0
+            return 0;
+        else
+        {
+            enum Sig = Bits!From-1;
+            static if(Bits!To < Bits!From)
+                return cast(To)(i >> (Sig-Bits!To));
+            else
+            {
+                To r;
+                enum numReps = Bits!To/Sig;
+                foreach(j; Iota!(1, numReps+1))
+                    r |= To(cast(Unsigned!From)(i&From.max)) << (Bits!To - j*Sig);
+                enum remain = Bits!To - numReps*Sig;
```

Comment:
Not sure what you mean, you want me to make the interior of the foreach enclosed in braces? What about blank line?